specs/required_block_state.md

```diff
@@ -0,0 +1,343 @@
+Record the block number (top of stack in the "BLOCKHASH" step), and the block hash (top
+of stack in the subsequent step).
+
+### `use_required_block_state`
+
+1. Obtain `RequiredBlockState`, for example by calling `eth_getRequiredBlockState`
+2. Use `compression_procedure` to decompress the `RequiredBlockState`
+3. `verify_required_block_state`
+4. `trace_block_locally`
+
+### `verify_required_block_state`
+
+Check block hashes are canonical such as a node or against an accumulator of canonical
+block hashes. Check merkle proofs in the required block state.
+
+### `trace_block_locally`
+
+Obtain a block (`eth_getBlockByNumber` JSON-RPC method) with transaction bodies. Use an EVM
+and load it with the `RequiredBlockState` and the block. Execute
+the transactions in the block and observe the trace.
+
+### `compression_procedure`
+
+The `RequiredBlockState` returned by the `eth_getRequiredBlockState` JSON-RPC method is
+compressed. Snappy compression is used ([https://github.com/google/snappy](https://github.com/google/snappy)).
+
+The encoding and decoding procedures are the same as that used in the Ethereum consensus specifications
+([https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#ssz-snappy-encoding-strategy](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#ssz-snappy-encoding-strategy)).
+
+For encoding (compression), data is first SSZ-encoded and then snappy-encoded.
+For decoding (decompression), data is first snappy-decoded and then SSZ-decoded.
+
+## Security
+
+### Future protocol changes
+
+Merkle patricia proofs may be replaced by verkle proofs after some hard fork.
+This would not invalidate `RequiredBlockState` data prior to that fork.
+The new proof format could be added to this specification for data after that fork.
+
+### Canonicality
+
+A recipient of `RequiredBlockState` must check that the blockhashes are part of the real
+Ethereum chain history. Failure to verify (`verify_required_block_state`) can result in invalid
+re-execution (`trace_block_locally`).
+
+### Post-block state root
+
+A user that has access to canonical block hashes and a sound EVM implementation has strong
+guarantees about the integrity of the block re-execution (`trace_block_locally`).
+
+However, there is no guarantee to be able to compute a new block state root for this post-execution
+state. For example, with the aim to check against the state root in the block header of that block
+and thereby audit the state changes that were applied.
+
+This is because the state changes may involve an arbitrary number of state deletions. State
+deletions may change the structure of the merkle trie in a way that requires knowledge of
+internal nodes that are not present in the proofs obtained by `eth_getProof` JSON-RPC method.
+Hence, while the complete post-block trie can sometimes be created, it is not guaranteed.
```

Comment:
Yup, I was worried about this.

Being able to consistently verify the post-state hash doesn't really feel optional to me. I can't trust the result of a local block execution that doesn't also prove that it got the same result as the network did. In the context of the Portal Network, nodes must validate data and only gossip data that we can prove locally.

EVM execution engines should be able to handle the case of missing trie nodes at the refactor, and returning the information needed to collect the data. They **must** handle the case, if they want to run against a partial trie database, like if they're running Beam Sync.

I am only familiar enough with py-evm to give the example:
https://github.com/ethereum/py-evm/blob/d751dc8c9c8199a16043a483b19c9f4d7a592202/eth/db/account.py#L605-L661

The `MissingTrieNode` exceptions are the relevant moments when the EVM realizes that it's missing some intermediate nodes that are required to prove the final state root, if you're only running with the state proof as defined in the current spec.

As a prototype, I suppose you could literally run py-evm with the unverified proofs, then retrieve the missing trie nodes over devp2p one by one (there usually aren't too many, from what I've seen).

Reply:
> .. then retrieve the missing trie nodes over devp2p one by one

For the missing trie node(s), I don't see a clear mechanism to obtain that node. For removed nodes that require knowledge of a sibling node, how could that sibling node be obtained? As it is mid-traversal, the terminal keys of the affected sibling are not trivially known. So `eth_getProof(block_number, key_involving_sibling_node)` cannot be used. A new method `get_trie_node_at_block_height(block_number, trie_node_hash)` could be written for a node, but this is nontrivial. More context here:
- https://github.com/perama-v/archors/issues/6#issuecomment-1598142177

Reply:
> .. nodes must validate data and only gossip data that we can prove locally.

This is preserved (gossiped data is validated). The gossiped data is the merkle proofs of the state, so this is validated, and the block is also validated. 

So the source of error here is a bug in the EVM implementation. 

```mermaid
flowchart TD
    Block[Block, secured by header: Program to run] -- gossiped --> Pre
    State[State, secured by merkle proofs: Input to program] -- gossiped --> Pre
    Pre[Program with inputs] -- Load into EVM environment -->  EVM[EVM executes]
    EVM -- bug here --> Post[Post block state]
    Post -.-> PostRoot[Post-block state root useful for these bugs]
    EVM -- bug here --> Trace[debug_traceBlock output]
```
I agree EVM bugs are possible and having the post-state root would be nice. 

However: 
- EVM impls are often shared across different client types, reducing bug risk. 
- Even having the post-block state does not guarantee that `debug_traceBlock` output is correct. It contains many
details that are not covered by a post-block state root. So one still needs to check that the EVM doesn't have bugs.
- Test suites can be run to compare debug_TraceBlock against the same result from an archive node (or different EVM implementation hooked into the portal network). This protects against EVM errors that result in bad state, and errors that result in bad traces.

Reply:
> Yup, I was worried about this.
> 
> Being able to consistently verify the post-state hash doesn't really feel optional to me. I can't trust the result of a local block execution that doesn't also prove that it got the same result as the network did. In the context of the Portal Network, nodes must validate data and only gossip data that we can prove locally.
> 
> EVM execution engines should be able to handle the case of missing trie nodes at the refactor, and returning the information needed to collect the data. They **must** handle the case, if they want to run against a partial trie database, like if they're running Beam Sync.
> 
> I am only familiar enough with py-evm to give the example: https://github.com/ethereum/py-evm/blob/d751dc8c9c8199a16043a483b19c9f4d7a592202/eth/db/account.py#L605-L661
> 
> The `MissingTrieNode` exceptions are the relevant moments when the EVM realizes that it's missing some intermediate nodes that are required to prove the final state root, if you're only running with the state proof as defined in the current spec.
> 
> As a prototype, I suppose you could literally run py-evm with the unverified proofs, then retrieve the missing trie nodes over devp2p one by one (there usually aren't too many, from what I've seen).

Ive proposed an[ idea here ](https://github.com/sogolmalek/EIP-x/blob/main/Readme.md) and believe our proposal on ZK proofs of the last state can indeed can help address the challenge mentioned. ZK proofs, specifically Zero-Knowledge Succinct Non-Interactive Argument of Knowledge (ZK-SNARKs), have the potential to provide proofs of complex computations and statements: https://github.com/sogolmalek/EIP-x/issues/6

Reply:
> ... our proposal on ZK proofs of the last state can indeed can help address the challenge mentioned

Let me see if I understand your proposal. The data that a peer receives could consist of the:
- (existing) The block pre-state (state required for the block, secured by merkle root in prior block and a header accumulator)
- (existing) The block (to allow the user to replay the block for any purpose)
- (new) A proof of block execution (ZK-EVM) consisting of a ZK-SNARK proof for the set of block post-state values (values that were accessed by the block). 

The user re-executes the block and arrives at post-block state values. Those are compared to the values in the ZK proof. If they are the same, the chance that the EVM and the ZK-EVM both having a bug is unlikely and the state transition is likely sound. 

So the ZK proof is equivalent to replaying the block using a different EVM implementation. That is, the same as getting the post-block state from a Rust implementation and comparing to the state produced by a Go implementation. 

In this respect, the presence of the ZK proof does not seem to introduce additional safety. Perhaps I overlooked a component?

Reply:
Here is one extension ("Exclusion proof helpers") to the `get_proofs` algorithm and `RequiredBlockState` data structure that may allow for the post-block state root to be computed. 

```md
### `get_proofs`

Call the `eth_getProof` JSON-RPC method for each state key (address) returned by the
`get_state_accesses` algorithm, including
storage keys if appropriate.

The block number (`n - 1`) used is the block prior to the block of interest (`n`) (state is stored as post-block
state).

For all account proofs, aggregate and sort the proof nodes and represent each proof as a list of
indices to those nodes. Repeat for all storage proofs.

*Extension: Exclusion proof helpers*: 
1. Obtain additional trie nodes for removed keys. 
2. Call `eth_getProof` again for each state key, this time for the block of interest (block `n`). 
3. For any key that has an inclusion proof in `n - 1`, and an  exclusion proof in block `n`, retain the proof for that key. 
4. Store these additional exclusion proofs in the `RequiredBlockState` data structure. 
5. When a user re-executes the block, if a key is removed in a way that requires additional knowledge 
6. of trie nodes to complete the trie, they can be sourced from the additional exclusion proofs. 
```
The problem with key deletion is that sometimes sibling nodes in the trie are required that aren't part of the `eth_getProof` data. The resolution is to identify those cases and use `eth_getProof` on the subsequent block and include that in the distributed data. 

The costs are:
- Additional `eth_getProof` calls (at time of construction). (2x number of calls)
- Some extra proofs in the distributed data (likely infrequent and small). (Estimate <5% increase in data size)

The benefits are:
- The post-block state root can be computed. This provides a valuable sanity check for the local EVM being used.
- No ZK proofs are required

Reply:
Thanks for your detailed arguments. I'm just trying to make sure i understand points well and hope i didnt go far wrong with my following arguments. love to learn more here. 

Our approach centers around the principle of **minimizing data reliance while ensuring the accuracy and reliability of state transitions**.While your understanding of our proposal is mostly accurate, there are a few points that could be considered when evaluating the introduced ZK proofs:
The process of re-executing the block on a different EVM implementation (Rust vs. Go, as you mentioned) can be cumbersome and resource-intensive.
another point is Attack Vectors and Determinism: While replaying the block using different EVM implementations is conceptually similar, it introduces the potential for inconsistencies between different implementations due to subtle differences in execution logic or determinism. furthermore, I believe ZK can ensure way more  Efficiency and Scaling. Re-executing the block using different EVM implementations **might be feasible on a small scale,** but it becomes much more challenging as the **network scales and the number of transactions increases.** ZK proofs, on the other hand, can be generated and verified more efficiently, making them more suitable for large-scale systems. 

As you've mentioned, the problem with key deletions is that sometimes sibling nodes in the trie are required to reconstruct the trie structure for verification purposes. These sibling nodes might not be available through the eth_getProof data, leading to incomplete trie reconstructions.

I think, a ZK proof of the last state involves generating a cryptographic proof that certifies the correctness of the entire state transition process, including deletions and modifications, can help. This proof can be designed to include information about sibling nodes that are required for verification. In essence, the ZK proof encapsulates the entire state transition, and by design, it must account for all the necessary data, including sibling nodes, to be valid.

Also By obtaining and validating a ZK proof of the last state, we can ensure that all required data for verifying the state transition, including missing sibling nodes, is included in the proof. This provides a complete and comprehensive validation mechanism that mitigates the challenge posed by missing nodes.

Reply:
Some additional context is that `eth_getRequiredBlockState` is designed to provide the minimum information required to re-execute an old block. The goal is to be able to download individual blocks and re-execute them in order to inspect every EVM step in that block. That is, the goal is to run the EVM (to gain historical insight).

With a ZK EVM, one can demonstrate to a peer that the block post-state is valid with respect to a block. This means that they do not have to re-execute the EVM themselves. This is very beneficial a light client that wants to trustlessly keep up to date with the latest Ethereum state. That is, the goal is to not run the EVM (to save resources).

Reply:
> For any key that has an inclusion proof in `n - 1`, and an  exclusion proof in block `n`, retain the proof for that key. 
4. Store these additional exclusion proofs in the `RequiredBlockState` data structure.

I think the idea is a good direction, but after discussing with @gballet I have the feeling sending the exclusion proofs themselves are not enough for the user to perform deletion locally. Because the sibling of the node being deleted could be a subtrie. So the client should use this approach (or another) to give the full prestate that is required for execution of the block.

Geth has the [prestateTracer](https://github.com/ethereum/go-ethereum/blob/2e02c1ffd9dffd1ec9e43c6b66f6c9bd1e556a0b/eth/tracers/native/prestate.go) which I noticed suffers from the same issue after reading up on this ticket. I'd be keen to fix it.

Reply:
At present this spec contains sufficient data to execute the block, but not update the proofs to get the post-block root. 

For context, the main issue is that sometimes the trie depth is changed, which impacts cousin-nodes. More on this here: https://github.com/perama-v/archors/blob/main/crates/multiproof/README.md

To enable that, I have been prototyping a mechanism that allows this using only available JSON-RPC methods (eth_getProof) without hacking an execution client. This method calls get_proof on the subsequent block gets these edge-case nodes and tacks them into RequiredBlockState as "oracle data". The data allows the trie update to complete and the post block root compute, verifying that the oracle data was in fact correct.

I learned that @s1na has taken a different approach, which is to modify Geth to record all trie nodes touched during a block execution (including post-root computation). That method is sufficient to get trie nodes required to compute the post-block root (but requires modifying the execution client). It is a nice approach.

Reply:
Essentially the oracle approach says: Look, some changes to the surrounding trie will happen, and they may be complex/extensive. One can "look in to the future" at this exact point in the traversal and just get the nodes for the affected part (leafy-end) of the traversal. The details of the surrounding trie don't need to be computed and aren't important because we can check that what we got from the future was in fact correct.

Bit of a hacky approach. I can see that this might be difficult to implement if not using eth_getProof.